### PR TITLE
proxy: Enabled dry mode (run with simulated nodes)

### PR DESCRIPTION
**Pull request by wprzytula**
It has appeared to be useful to have a feature in the proxy to interact with the driver without having any Scylla nodes up. This is especially useful for unit testing.

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] ~~I added appropriate `Fixes:` annotations to PR description.~~

### CODE REVIEW

**Comment by piodul** [comment]
Could you provide some examples where dry mode would be useful? Currently, we require a Scylla cluster to be available when running `cargo test`, so I don't understand why we need the dry mode.

**Comment by wprzytula** [comment]
> Could you provide some examples where dry mode would be useful? Currently, we require a Scylla cluster to be available when running `cargo test`, so I don't understand why we need the dry mode.

Consider a situation that you want to test particular behaviour in case of Scylla or Cassandra in versions that are different than those available in the test cluster. Then you can simply mock their behaviour with the proxy running in dry mode.

**Comment by piodul** [comment]
> Consider a situation that you want to test particular behaviour in case of Scylla or Cassandra in versions that are different than those available in the test cluster. Then you can simply mock their behaviour with the proxy running in dry mode.

Could you give some examples of the version-specific behaviors that you would like to test in this manner?

If we want to support multiple versions of Scylla/Cassandra, we should just run all the tests multiple times, against all supported versions.

**Comment by wprzytula** [comment]
> > Consider a situation that you want to test particular behaviour in case of Scylla or Cassandra in versions that are different than those available in the test cluster. Then you can simply mock their behaviour with the proxy running in dry mode.
> 
> Could you give some examples of the version-specific behaviors that you would like to test in this manner?
> 
> If we want to support multiple versions of Scylla/Cassandra, we should just run all the tests multiple times, against all supported versions.

In this PR, I'm testing various server-side behaviour regarding LWT_OPTIMIZATION_MARK. In this exact case, I could do with running one test on Cassandra workflow and another one on Scylla, but in general it is often much more precise to test some feature in isolation, when the environment (in this case the server side) is strictly controlled. That's what the proxy's dry mode is for.

**Review by piodul — CHANGES_REQUESTED** [review]
Please split the PR into smaller steps. There is a lot of things going on at once and it's very hard to track all of the changes if they are all introduced in one commit.

**Comment by wprzytula** [comment]
> Please split the PR into smaller steps. There is a lot of things going on at once and it's very hard to track all of the changes if they are all introduced in one commit.

I've done my best and extracted two more commits: removal of a small piece of dead code and a unit test. However, the vast majority of this PR must be given in a single commit, as it's an encapsulated unity - the code would not compile if I split it in any way. Will you accept it after this explanation?

**Comment by piodul** [comment]
> I've done my best and extracted two more commits: removal of a small piece of dead code and a unit test. However, the vast majority of this PR must be given in a single commit, as it's an encapsulated unity - the code would not compile if I split it in any way. Will you accept it after this explanation?

I still think the commit `proxy: Enabled dry mode (run with simulated nodes)` is much too complicated. A good, readable PR is not just about taking the final diff and splitting it into multiple commits. Ideally, every commit should do a single, logical step. For example, refactors and adding functionality should be separate.

I don't think I agree that the majority of the PR must be done in a single commit. I can see that the problematic commit does multiple things and you could try to introduce them step by step - here are some examples:

- You are moving some fields from `Node` to `NodeType::Real`. You could introduce `NodeType` with a single variant `NodeType::Real` and move the fields there in one commit.
- The same with conversion of `InternalNode` to an enum: `struct InternalNode` looks like `InternalNode::Real`, so you could move the fields in one commit.
- `response_rules` have been wrapped in an `Option` - maybe could be moved to its own commit?
- It looks like you moved a lot of code to `spawn_workers` - could this movement be a separate commit?
- Adding `NodeType::Simulated` and `InternalNode::Simulated` - could probably be done in separate or a single commit, whatever makes more sense.

I don't see more of them at the moment (but most likely there are more), and I may be wrong about some of them because I didn't dig too deeply into the commit.

Mixing multiple refactors with adding features results in commits which are hard to read and whose changes are hard to follow. As a reviewer, I would like to verify that those things were done correctly, but it's hard to do when they are interleaved.

Please remember to describe the reason for each commit in their messages. This includes refactors as well.

**Comment by wprzytula** [comment]
> I still think the commit `proxy: Enabled dry mode (run with simulated nodes)` is much too complicated. A good, readable PR is not just about taking the final diff and splitting it into multiple commits. Ideally, every commit should do a single, logical step. For example, refactors and adding functionality should be separate.

Thank you for that detailed advice. I followed it and the refactored commits are ready for review. I hope you will grasp it easily now.

File: scylla-proxy/src/proxy.rs [hunk 1]
```diff
@@ -61,12 +61,17 @@ impl ShardAwareness {
     }
 }
 
+/// Node can be either Real (truly backed by a Scylla node) or Simulated
```
Comment (piodul) [inline scylla-proxy/src/proxy.rs:64]: Nit: I think it would be actually fine to merge `proxy: Added Node::Simulated for dry mode` commit with the one that introduces the `NodeType::Simulated` enum variant - they are logically related to each other and one wouldn't obscure the other. Besides, you could avoid having to use `#[allow(unused)]`.

File: scylla-proxy/src/proxy.rs [hunk 2]
```diff
@@ -419,24 +456,27 @@ impl Doorkeeper {
...
             .map_err(|err| DoorkeeperError::DriverConnectionAttempt(node.proxy_addr(), err))?;
 
-        #[allow(irrefutable_let_patterns)]
         if let InternalNode::Real {
             shard_awareness,
             real_addr,
             ..
         } = node
         {
-            let doorkeeper_type = if shard_awareness.is_aware() {
```
Comment (piodul) [inline scylla-proxy/src/proxy.rs]: This is an unrelated refactor. Please move it to a separate commit.

**Review by piodul — COMMENTED** [review]
Thanks, it looks much better now. I think some commits could still be split a bit further and I left some suggestions, but I also tried to leave some comments regarding implementation itself.

One more thing that I would appreciate is a short, high-level description of the dry mode, i.e. what it is, what for, how it works and how it affects the operation of the proxy. For example, I remember that during review of the LWT PR it wasn't obvious for me that the proxy just ignores the requests that aren't handled by proxy's rules. This description should be placed in the PR, and - even better - as a docstring in the code, so that other people that will work on the proxy in the future will have some high-level overview.

File: scylla-proxy/src/proxy.rs [hunk 3]
```diff
@@ -147,6 +147,25 @@ struct InternalNode {
     request_rules: Arc<Mutex<Vec<RequestRule>>>,
     response_rules: Arc<Mutex<Vec<ResponseRule>>>,
 }
+
+impl From<Node> for InternalNode {
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-proxy/src/proxy.rs:240]: I think it is too terse. I would rephrase it using simpler grammar and maybe elaborate on why you think this refactor enhances clarity.

For example: "The logic in Proxy::new responsible for converting Node to InternalNode is abstracted away into `impl From<Node> for InternalNode`. This shortens and simplifies the Proxy::new function."

**Comment by wprzytula** [comment]
WDYT now, @piodul?

File: scylla-proxy/src/proxy.rs [hunk 4]
```diff
@@ -1933,4 +1933,132 @@ mod tests {
...
+        };
+        let opcode = FrameOpcode::Request(RequestOpcode::Options);
+
+        let body = random_body();
+
+        let mut conn = TcpStream::connect(node1_proxy_addr).await.unwrap();
+
+        write_frame(params, opcode, &body, &mut conn).await.unwrap();
+        // We assert that after sufficiently long time, no error happens inside proxy.
+        tokio::time::sleep(Duration::from_millis(3)).await;
```
Comment (piodul) [inline scylla-proxy/src/proxy.rs:1957]: Note for the future: we should have a way to gracefully close the proxy, i.e. prevent it from accepting more connections and wait until the current connections are closed by the client. This would allow us to avoid such sleeps as above.

File: scylla-proxy/src/proxy.rs (hunk 2 above)
Comment (piodul) [reply scylla-proxy/src/proxy.rs]: Not really resolved, but I'll let it slide this time ;)

**Review by piodul — APPROVED** [review]
LGTM

Thanks for taking the time to split everything up into parts, it made the review considerably easier.
